File: frontend/src/engine/simulation.js
/**
 * engine/simulation.js
 * WH40K 10e Monte Carlo simulation engine — port of backend/engine/simulation.py
 *
 * Entry point: simulate(req) -> SimResult
 *
 * req shape:
 * {
 *   attacker: { models, weapon: { name, attacks, skill, strength, ap, damage, keywords }, buffs },
 *   defender: { toughness, save, invuln, wounds, models, fnp, keywords },
 *   context:  { cover, half_range, attacker_moved, attacker_charged, target_visible },
 *   n_trials: number
 * }
 */

import { roll, clamp, woundThreshold, d6 } from './dice.js'

// ── Keyword / buff helpers ────────────────────────────────────────────────────

function kw(keywords, type) {
  return keywords.find((k) => k.type === type) ?? null
}

function hasKw(keywords, type) {
  return keywords.some((k) => k.type === type)
}

function buf(buffs, type) {
  return buffs.find((b) => b.type === type) ?? null
}

function bufVal(buffs, type, defaultVal = 0) {
  const b = buf(buffs, type)
  return b ? b.value : defaultVal
}

// ── Single trial ──────────────────────────────────────────────────────────────

function simulateOnce(req) {
  const w   = req.attacker.weapon
  const d   = req.defender
  const ctx = req.context
  const kws = w.keywords ?? []
  const bufs = req.attacker.buffs ?? []

  // Keyword flags
  const hasTorrent   = hasKw(kws, 'TORRENT')
  const hasLethal    = hasKw(kws, 'LETHAL_HITS')
  const hasSustained = hasKw(kws, 'SUSTAINED_HITS')
  const hasDevWounds = hasKw(kws, 'DEVASTATING_WOUNDS')
  const hasBlast     = hasKw(kws, 'BLAST')
  const hasHeavy     = hasKw(kws, 'HEAVY')
  const hasLance     = hasKw(kws, 'LANCE')
  const hasTwin      = hasKw(kws, 'TWIN_LINKED')
  const hasRapid     = hasKw(kws, 'RAPID_FIRE')
  const hasMelta     = hasKw(kws, 'MELTA')
  const hasIgCover   = hasKw(kws, 'IGNORES_COVER')
  const hasIndirect  = hasKw(kws, 'INDIRECT_FIRE')
  const hasExtraAtk  = hasKw(kws, 'EXTRA_ATTACKS')

  const kwSustained = kw(kws, 'SUSTAINED_HITS')
  const kwRapid     = kw(kws, 'RAPID_FIRE')
  const kwMelta     = kw(kws, 'MELTA')
  const kwExtraAtk  = kw(kws, 'EXTRA_ATTACKS')
  const kwAnti      = kw(kws, 'ANTI')
  const kwCritHit   = kw(kws, 'CRITICAL_HIT_ON')

  const bufCritHit   = buf(bufs, 'CRITICAL_HIT_ON')
  const bufCritWound = buf(bufs, 'CRITICAL_WOUND_ON')

  const critHitThr   = Math.min(
    kwCritHit  ? kwCritHit.value  : 6,
    bufCritHit ? bufCritHit.value : 6,
  )
  const critWoundThr = bufCritWound ? bufCritWound.value : 6

  // Stat modifiers
  const atkMod   = bufVal(bufs, 'ATTACKS_MODIFIER')
  const strMod   = bufVal(bufs, 'STRENGTH_MODIFIER')
  const apMod    = bufVal(bufs, 'AP_MODIFIER')
  const dmgMod   = bufVal(bufs, 'DAMAGE_MODIFIER')
  let   hitMod   = bufVal(bufs, 'HIT_MODIFIER')
  let   woundMod = bufVal(bufs, 'WOUND_MODIFIER')
  const saveMod  = bufVal(bufs, 'SAVE_MODIFIER')

  const rerollHits   = buf(bufs, 'REROLL_HITS')
  const rerollWounds = buf(bufs, 'REROLL_WOUNDS')
  const rerollSaves  = buf(bufs, 'REROLL_SAVES')

  const effStrength = w.strength + strMod
  const effAp       = w.ap + apMod

  // Context modifiers
  if (hasHeavy && !ctx.attacker_moved)  hitMod += 1
  if (hasLance && ctx.attacker_charged) woundMod += 1
  if (hasIndirect && !ctx.target_visible) hitMod -= 1

  hitMod   = clamp(hitMod,   -1, 1)
  woundMod = clamp(woundMod, -1, 1)

  // ── Phase 1: attacks ─────────────────────────────────────────────────────────

  const basePerModel  = Math.max(1, roll(w.attacks) + atkMod)
  const extraPerModel = hasExtraAtk ? roll(kwExtraAtk.value) : 0
  const blastBonus    = hasBlast ? Math.floor(d.models / 5) : 0

  let numAttacks = req.attacker.models * (basePerModel + extraPerModel) + blastBonus

  if (hasRapid && ctx.half_range) {
    numAttacks += roll(kwRapid.value) * req.attacker.models
  }

  // ── Phase 2: hit rolls ───────────────────────────────────────────────────────

  let autoWounds = 0
  let hits = 0

  if (hasTorrent) {
    hits = numAttacks
  } else {
    const skill = clamp(w.skill, 2, 6)

    for (let i = 0; i < numAttacks; i++) {
      let die = d6()

      if (rerollHits) {
        const eligible =
          (rerollHits.value === 'ones' && die === 1) ||
          (rerollHits.value === 'all'  && clamp(die + hitMod, 1, 6) < skill && die < critHitThr)
        if (eligible) die = d6()
      }

      if (die === 1) continue

      const isCrit   = die >= critHitThr
      const modified = clamp(die + hitMod, 1, 6)
      const success  = isCrit || modified >= skill

      if (!success) continue

      hits++

      if (isCrit) {
        if (hasLethal)    autoWounds += 1
        if (hasSustained) hits += roll(kwSustained.value)
      }
    }
  }

  // ── Phase 3: wound rolls ─────────────────────────────────────────────────────

  const wThr = woundThreshold(effStrength, d.toughness)
  const hitsToRoll = Math.max(0, hits - autoWounds)

  let woundsNormal = 0
  let mortalWounds = 0

  const canRerollWound = hasTwin || rerollWounds !== null

  function rerollWoundEligible(die) {
    if (hasTwin) return true
    if (rerollWounds) {
      return (rerollWounds.value === 'ones' && die === 1) || rerollWounds.value === 'all'
    }
    return false
  }

  function evaluateWound(dVal) {
    if (dVal === 1) return { success: false, isCrit: false }
    let isCrit = dVal >= critWoundThr
    if (kwAnti && dVal >= kwAnti.threshold) {
      const defKws = (d.keywords ?? []).map((k) => k.toUpperCase())
      if (defKws.includes(kwAnti.target.toUpperCase())) isCrit = true
    }
    const modified = clamp(dVal + woundMod, 1, 6)
    const success  = isCrit || modified >= wThr
    return { success, isCrit }
  }

  for (let i = 0; i < hitsToRoll; i++) {
    let die = d6()
    let { success, isCrit } = evaluateWound(die)

    if (!success && canRerollWound && rerollWoundEligible(die)) {
      die = d6()
      ;({ success, isCrit } = evaluateWound(die))
    }

    if (!success) continue

    if (isCrit && hasDevWounds) {
      mortalWounds++
    } else {
      woundsNormal++
    }
  }

  const woundsNeedingSave = woundsNormal + autoWounds

  // ── Phase 4: save rolls ──────────────────────────────────────────────────────

  // AP is stored as 0, -1, -2 etc. save worsens by |AP|
  // save=3, ap=-2 → armor_sv = 3 - (-2) = 5+ (worse)
  let armorSv = d.save - effAp

  if (ctx.cover && !hasIgCover) armorSv -= 1  // cover improves by 1

  armorSv = Math.max(armorSv, 2)

  let effSave = d.invuln != null ? Math.min(armorSv, d.invuln) : armorSv
  effSave = Math.min(effSave, 7)
  effSave += saveMod

  let failedSaves = 0

  for (let i = 0; i < woundsNeedingSave; i++) {
    let die = d6()

    if (rerollSaves) {
      const eligible =
        (rerollSaves.value === 'ones' && die === 1) ||
        (rerollSaves.value === 'all'  && die < effSave)
      if (eligible) die = d6()
    }

    if (die < effSave) failedSaves++
  }

  const totalUnsaved = failedSaves + mortalWounds

  // ── Phase 5: damage + FNP ────────────────────────────────────────────────────

  let totalDamage = 0

  for (let i = 0; i < totalUnsaved; i++) {
    let dmg = roll(w.damage) + dmgMod

    if (hasMelta && ctx.half_range) dmg += roll(kwMelta.value)

    dmg = Math.max(0, dmg)

    if (d.fnp != null) {
      for (let j = 0; j < dmg; j++) {
        if (d6() < d.fnp) totalDamage++  // fails FNP → damage lands
      }
    } else {
      totalDamage += dmg
    }
  }

  return totalDamage
}

// ── Aggregation ───────────────────────────────────────────────────────────────

export function simulate(req) {
  const n = req.n_trials ?? 1000
  const trials = []
  for (let i = 0; i < n; i++) trials.push(simulateOnce(req))

  const sorted = [...trials].sort((a, b) => a - b)
  const mean   = trials.reduce((s, v) => s + v, 0) / n
  const median = sorted[Math.floor(n / 2)]
  const std    = Math.sqrt(trials.reduce((s, v) => s + (v - mean) ** 2, 0) / Math.max(n - 1, 1))

  function pct(p) {
    const idx = Math.floor((p / 100) * n)
    return sorted[Math.min(idx, n - 1)]
  }

  // Histogram
  const counter = {}
  for (const v of trials) counter[v] = (counter[v] ?? 0) + 1
  const maxDmg = trials.length ? Math.max(...trials) : 0
  const histogram = []
  for (let dmg = 0; dmg <= maxDmg; dmg++) {
    const count = counter[dmg] ?? 0
    histogram.push({ damage: dmg, count, probability: round4(count / n) })
  }

  // Kill probabilities
  const woundsPerModel = req.defender.wounds
  const killProbs = {}
  for (let k = 1; k <= req.defender.models; k++) {
    const threshold = k * woundsPerModel
    killProbs[String(k)] = round4(trials.filter((t) => t >= threshold).length / n)
  }

  return {
    summary: {
      mean_damage:        round2(mean),
      median_damage:      round2(median),
      std_dev:            round2(std),
      p10:                pct(10),
      p25:                pct(25),
      p75:                pct(75),
      p90:                pct(90),
      mean_models_killed: round2(woundsPerModel > 0 ? mean / woundsPerModel : 0),
    },
    damage_histogram:   histogram,
    kill_probabilities: killProbs,
    n_trials:           n,
  }
}

function round2(v) { return Math.round(v * 100) / 100 }
function round4(v) { return Math.round(v * 10000) / 10000 }

File: backend/engine/simulation.py
"""
engine/simulation.py
--------------------
WH40K 10e Monte Carlo simulation engine.
Pure Python — no FastAPI, no database, no I/O.

Entry point: simulate(req: SimRequest) -> SimResult
"""

import math
import random
import statistics
from collections import Counter

from .dice import clamp, roll, wound_threshold
from .schemas import (
    SimAttacker, SimBuff, SimContext, SimDefender, SimRequest, SimResult,
    SimSummary, DamageBucket, WeaponKeyword,
)


# ── Helpers ───────────────────────────────────────────────────────────────────

def _kw(keywords: list[WeaponKeyword], kw_type: str):
    """Return first keyword of given type, or None."""
    for k in keywords:
        if k.type == kw_type:
            return k
    return None


def _has_kw(keywords: list[WeaponKeyword], kw_type: str) -> bool:
    return any(k.type == kw_type for k in keywords)


def _buf(buffs: list[SimBuff], buf_type: str):
    """Return first buff of given type, or None."""
    for b in buffs:
        if b.type == buf_type:
            return b
    return None


def _buf_val(buffs: list[SimBuff], buf_type: str, default: int = 0) -> int:
    b = _buf(buffs, buf_type)
    return b.value if b else default


# ── Single trial ──────────────────────────────────────────────────────────────

def _simulate_once(req: SimRequest) -> int:
    """Run one Monte Carlo trial. Returns total damage dealt to the defender."""

    w    = req.attacker.weapon
    d    = req.defender
    ctx  = req.context
    kws  = w.keywords
    bufs = req.attacker.buffs

    # -- Keyword presence flags --
    has_torrent    = _has_kw(kws, "TORRENT")
    has_lethal     = _has_kw(kws, "LETHAL_HITS")
    has_sustained  = _has_kw(kws, "SUSTAINED_HITS")
    has_devwounds  = _has_kw(kws, "DEVASTATING_WOUNDS")
    has_blast      = _has_kw(kws, "BLAST")
    has_heavy      = _has_kw(kws, "HEAVY")
    has_lance      = _has_kw(kws, "LANCE")
    has_twin       = _has_kw(kws, "TWIN_LINKED")
    has_rapid      = _has_kw(kws, "RAPID_FIRE")
    has_melta      = _has_kw(kws, "MELTA")
    has_igcover    = _has_kw(kws, "IGNORES_COVER")
    has_indirect   = _has_kw(kws, "INDIRECT_FIRE")
    has_extra_atk  = _has_kw(kws, "EXTRA_ATTACKS")

    kw_sustained  = _kw(kws, "SUSTAINED_HITS")
    kw_rapid      = _kw(kws, "RAPID_FIRE")
    kw_melta      = _kw(kws, "MELTA")
    kw_extra_atk  = _kw(kws, "EXTRA_ATTACKS")
    kw_anti       = _kw(kws, "ANTI")
    kw_crit_hit   = _kw(kws, "CRITICAL_HIT_ON")

    # -- Critical thresholds (take the most aggressive: lowest value) --
    buf_crit_hit   = _buf(bufs, "CRITICAL_HIT_ON")
    buf_crit_wound = _buf(bufs, "CRITICAL_WOUND_ON")

    crit_hit_thr = min(
        kw_crit_hit.value   if kw_crit_hit   else 6,
        buf_crit_hit.value  if buf_crit_hit  else 6,
    )
    crit_wound_thr = buf_crit_wound.value if buf_crit_wound else 6

    # -- Stat modifiers --
    atk_mod   = _buf_val(bufs, "ATTACKS_MODIFIER")
    str_mod   = _buf_val(bufs, "STRENGTH_MODIFIER")
    ap_mod    = _buf_val(bufs, "AP_MODIFIER")
    dmg_mod   = _buf_val(bufs, "DAMAGE_MODIFIER")
    hit_mod   = _buf_val(bufs, "HIT_MODIFIER")
    wound_mod = _buf_val(bufs, "WOUND_MODIFIER")
    save_mod  = _buf_val(bufs, "SAVE_MODIFIER")

    reroll_hits   = _buf(bufs, "REROLL_HITS")
    reroll_wounds = _buf(bufs, "REROLL_WOUNDS")
    reroll_saves  = _buf(bufs, "REROLL_SAVES")

    # -- Effective weapon stats --
    eff_strength = w.strength + str_mod
    eff_ap       = w.ap + ap_mod

    # Context-driven modifiers
    if has_heavy and not ctx.attacker_moved:
        hit_mod += 1
    if has_lance and ctx.attacker_charged:
        wound_mod += 1
    if has_indirect and not ctx.target_visible:
        hit_mod -= 1

    # WH40K 10e: roll modifiers are capped at ±1
    hit_mod   = clamp(hit_mod,   -1, 1)
    wound_mod = clamp(wound_mod, -1, 1)

    # ── Phase 1: number of attacks ────────────────────────────────────────────

    # Base attacks per model (affected by ATTACKS_MODIFIER)
    base_per_model = max(1, roll(w.attacks) + atk_mod)

    # EXTRA_ATTACKS: bonus attacks not affected by ATTACKS_MODIFIER
    extra_per_model = roll(kw_extra_atk.value) if has_extra_atk else 0

    # BLAST: +floor(N/5) attacks where N = defender models
    blast_bonus = math.floor(d.models / 5) if has_blast else 0

    num_attacks = req.attacker.models * (base_per_model + extra_per_model) + blast_bonus

    # RAPID FIRE: +value attacks per model if within half range
    if has_rapid and ctx.half_range:
        num_attacks += roll(kw_rapid.value) * req.attacker.models

    # ── Phase 2: hit rolls ────────────────────────────────────────────────────

    auto_wounds = 0   # LETHAL_HITS crits — skip wound roll, still need save
    hits = 0

    if has_torrent:
        hits = num_attacks
        # TORRENT has no actual dice rolls — no crits, no LETHAL/SUSTAINED triggers
    else:
        skill = clamp(w.skill, 2, 6)

        for _ in range(num_attacks):
            die = random.randint(1, 6)

            # Reroll before evaluation (natural die decides eligibility)
            if reroll_hits:
                eligible = (
                    (reroll_hits.value == "ones" and die == 1) or
                    (reroll_hits.value == "all"  and clamp(die + hit_mod, 1, 6) < skill and not (die >= crit_hit_thr))
                )
                if eligible:
                    die = random.randint(1, 6)

            if die == 1:
                continue  # natural 1 always fails, even after reroll

            is_crit = die >= crit_hit_thr
            modified = clamp(die + hit_mod, 1, 6)
            success = is_crit or (modified >= skill)

            if not success:
                continue

            hits += 1

            if is_crit:
                if has_lethal:
                    auto_wounds += 1       # wound is automatic — save still required
                if has_sustained:
                    hits += roll(kw_sustained.value)  # bonus hits, not critical

    # ── Phase 3: wound rolls ──────────────────────────────────────────────────

    w_thr = wound_threshold(eff_strength, d.toughness)

    # Hits that need a wound roll = total hits minus LETHAL auto-wounds
    hits_to_roll = max(0, hits - auto_wounds)

    wounds_normal  = 0  # need a save roll
    mortal_wounds  = 0  # DEVASTATING_WOUNDS: bypass all saves

    # Whether a reroll is available for wounds (TWIN_LINKED or buff — only one)
    can_reroll_wound = has_twin or (reroll_wounds is not None)

    def _reroll_wound_eligible(die: int, value: str | None) -> bool:
        """True if the die qualifies for reroll under the active rule."""
        if has_twin:
            # TWIN_LINKED re-rolls all failed wounds
            return True
        if reroll_wounds:
            return (
                (reroll_wounds.value == "ones" and die == 1) or
                reroll_wounds.value == "all"
            )
        return False

    for _ in range(hits_to_roll):
        die = random.randint(1, 6)

        def _evaluate_wound(d_val: int):
            is_crit = d_val >= crit_wound_thr
            # ANTI keyword: critical wound on X+ vs matching keyword
            if kw_anti and d_val >= kw_anti.threshold:
                if kw_anti.target.upper() in (k.upper() for k in d.keywords):
                    is_crit = True
            if d_val == 1:
                return False, False
            modified = clamp(d_val + wound_mod, 1, 6)
            success = is_crit or (modified >= w_thr)
            return success, is_crit

        success, is_crit = _evaluate_wound(die)

        # One reroll if eligible and failed
        if not success and can_reroll_wound and _reroll_wound_eligible(die, None):
            die = random.randint(1, 6)
            success, is_crit = _evaluate_wound(die)

        if not success:
            continue

        if is_crit and has_devwounds:
            mortal_wounds += 1   # bypasses armor save AND invuln
        else:
            wounds_normal += 1

    # Wounds needing a save = normal wounds + LETHAL auto-wounds
    wounds_needing_save = wounds_normal + auto_wounds

    # ── Phase 4: save rolls ───────────────────────────────────────────────────

    # Armor save degraded by AP (lower number = easier to save, but AP is negative)
    # effective_ap is <= 0, so: armor_sv = save + effective_ap (better save)
    # Example: save=3, ap=-2 → armor_sv = 3 + (-2) = 1... wait that's wrong
    # Actually: AP -2 means the save is WORSE by 2. save=3+, with AP-2 → save on 5+
    # So: armor_sv = save - ap (ap is stored as negative, so -(-2) = +2 → 3+2=5)
    armor_sv = d.save - eff_ap   # eff_ap is 0, -1, -2 etc. → -(-2) = +2

    # Cover gives +1 to armor save (makes it 1 easier = lower threshold)
    if ctx.cover and not has_igcover:
        armor_sv -= 1   # 5+ becomes 4+ (lower number = better save)

    armor_sv = max(armor_sv, 2)   # 2+ is the best possible armor save

    # Use best available save (armor or invuln — lower = better)
    if d.invuln is not None:
        eff_save = min(armor_sv, d.invuln)
    else:
        eff_save = armor_sv

    # 7+ is impossible (all dice show 1-6)
    eff_save = min(eff_save, 7)

    # Save modifier (from defender buff) — positive = harder to save
    eff_save += save_mod

    failed_saves = 0
    for _ in range(wounds_needing_save):
        die = random.randint(1, 6)

        if reroll_saves:
            eligible = (
                (reroll_saves.value == "ones" and die == 1) or
                (reroll_saves.value == "all"  and die < eff_save)
            )
            if eligible:
                die = random.randint(1, 6)

        if die < eff_save:
            failed_saves += 1

    # DEVASTATING_WOUNDS bypass all saves (mortal wounds)
    total_unsaved = failed_saves + mortal_wounds

    # ── Phase 5: damage + FNP ────────────────────────────────────────────────

    total_damage = 0

    for _ in range(total_unsaved):
        dmg = roll(w.damage) + dmg_mod

        # MELTA bonus if within half range
        if has_melta and ctx.half_range:
            dmg += roll(kw_melta.value)

        dmg = max(0, dmg)

        # Feel No Pain: roll per damage point — save on fnp+
        if d.fnp is not None:
            for _ in range(dmg):
                if random.randint(1, 6) < d.fnp:   # fails FNP → damage lands
                    total_damage += 1
        else:
            total_damage += dmg

    return total_damage


# ── Aggregation ───────────────────────────────────────────────────────────────

def simulate(req: SimRequest) -> SimResult:
    """Run n_trials simulations and return aggregated results."""

    trials = [_simulate_once(req) for _ in range(req.n_trials)]
    n = len(trials)
    sorted_t = sorted(trials)
    counter = Counter(trials)

    mean_dmg   = statistics.mean(trials)
    median_dmg = float(statistics.median(trials))
    std_dev    = statistics.stdev(trials) if n > 1 else 0.0

    def pct(p: float) -> float:
        idx = int(p / 100 * n)
        return float(sorted_t[min(idx, n - 1)])

    wounds_per_model = req.defender.wounds
    mean_killed = mean_dmg / wounds_per_model if wounds_per_model > 0 else 0.0

    max_dmg = max(trials) if trials else 0
    histogram = [
        DamageBucket(
            damage=dmg,
            count=counter.get(dmg, 0),
            probability=round(counter.get(dmg, 0) / n, 4),
        )
        for dmg in range(0, max_dmg + 1)
    ]

    # P(kills >= k models) for k in 1..defender.models
    kill_probs: dict[str, float] = {}
    for k in range(1, req.defender.models + 1):
        threshold = k * wounds_per_model
        prob = sum(1 for t in trials if t >= threshold) / n
        kill_probs[str(k)] = round(prob, 4)

    return SimResult(
        summary=SimSummary(
            mean_damage=round(mean_dmg, 2),
            median_damage=round(median_dmg, 2),
            std_dev=round(std_dev, 2),
            p10=pct(10),
            p25=pct(25),
            p75=pct(75),
            p90=pct(90),
            mean_models_killed=round(mean_killed, 2),
        ),
        damage_histogram=histogram,
        kill_probabilities=kill_probs,
        n_trials=n,
    )
